src/transformers/modeling_rope_utils.py

```diff
@@ -0,0 +1,481 @@
+    if not isinstance(long_factor, list) and all(isinstance(x, (int, float)) for x in long_factor):
+        raise ValueError(f"`rope_scaling`'s long_factor field must be a list of numbers, got {long_factor}")
+    if not len(long_factor) == dim // 2:
+        raise ValueError(f"`rope_scaling`'s long_factor field must have length {dim // 2}, got {len(long_factor)}")
+
+    # Handle Phi3 divergence: prefer the use of `attention_factor` and/or `factor` over
+    # `original_max_position_embeddings` to compute internal variables. The latter lives outside `rope_scaling` and is
+    # unique to longrope (= undesirable)
+    if hasattr(config, "original_max_position_embeddings"):
+        logger.warning_once(
+            "This model has set a `original_max_position_embeddings` field, to be used together with "
+            "`max_position_embeddings` to determine a scaling factor. Please set the `factor` field of `rope_scaling`"
+            "with this ratio instead -- we recommend the use of this field over `original_max_position_embeddings`, "
+            "as it is compatible with most model architectures."
+        )
+    else:
+        factor = rope_scaling.get("factor")
+        if factor is None:
+            raise ValueError("Missing required keys in `rope_scaling`: 'factor'")
+        elif not isinstance(factor, float) or factor < 1.0:
+            raise ValueError(f"`rope_scaling`'s factor field must be a float >= 1, got {factor}")
+
+        attention_factor = rope_scaling.get("attention_factor")
+        if attention_factor is not None and not isinstance(attention_factor, float) or attention_factor < 0:
+            raise ValueError(
+                f"`rope_scaling`'s attention_factor field must be a float greater than 0, got {attention_factor}"
+            )
+
+
+# Like `ROPE_INIT_FUNCTIONS`, this validation function mapping can be dynamically updated for custom RoPE types.
+ROPE_VALIDATION_FUNCTIONS = {
+    "default": _validate_default_rope_parameters,
+    "linear": _validate_linear_scaling_rope_parameters,
+    "dynamic": _validate_linear_scaling_rope_parameters,  # `dynamic` has the same validation pattern as `linear`
+    "yarn": _validate_yarn_parameters,
+    "longrope": _validate_longrope_parameters,
+}
+
+
+def rope_config_validation(config: PretrainedConfig):
+    """
+    Validate the RoPE config arguments, given a `PretrainedConfig` object
+    """
+    rope_scaling = config.rope_scaling
+    if rope_scaling is None:
+        return
+
+    possible_rope_types = set(ROPE_INIT_FUNCTIONS.keys())
+    rope_type = rope_scaling.get("type")
+    if rope_type is None:
+        raise ValueError(
+            f"rope_scaling must contain a non-None 'type' field. Possible options are {possible_rope_types}"
+        )
+
+    validation_fn = ROPE_VALIDATION_FUNCTIONS.get(rope_type)
+    if validation_fn is not None:
+        validation_fn(config)
+    # else: no validation, it is a registered custom RoPE type without validation
```

Comment:
Is this something we expect users to do / something we do in the repo ourselves? 

If it's something we expect users to do, then all of the `compute` methods should be public. Having them as private implies all of the rope logic should be hidden from the user.

If it's something we do, then I think we should always have validation set up. In fact, is there a reason to not enforce validation, even for custom extensions? 

If we decide to allow no validation, we should emit a `logger.info` or `logger.debug` message indicating that no validation is happening.

Reply:
changed to force validation (sorry, this was copied from a previous iteration where it made more sense 😅 )